lib/registry-secrets.ts

```diff
@@ -0,0 +1,94 @@
+/**
+ * @license
+ * Copyright 2018 Balena
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * This module contains interfaces and classes for the validation of the JSON
+ * format expected by the balena builder and the docker daemon for the
+ * authentication of private docker registries.
+ */
+
+import * as ajv from 'ajv';
+
+import { RegistrySecretValidationError } from './errors';
+
+export interface RegistrySecrets {
+	[registryAddress: string]: {
+		username: string;
+		password: string;
+	};
+}
+
+/**
+ * JSON schema validator for the private registry "secrets" (username and
+ * password entries keyed by hostname:port registry addresses).
+ */
+export class RegistrySecretValidator {
+	private registrySecretJsonSchema = {
+		// Sample valid registrySecrets JSON contents:
+		//   {  "docker.example.com": {"username": "ann", "password": "hunter2"},
+		//      "https://idx.docker.io/v1/": {"username": "mck", "password": "cze14"}
+		//   }
+		type: 'object',
+		patternProperties: {
+			'^\\S+$': {
```

Comment:
This file was moved here from [another PR](https://github.com/balena-io/resin-builder/pull/562). The pattern `'^\\S+$'` used to be `'.+'` and Cameron had added a comment: _"Do we want anything slightly more stringent here?"_  I guess this is sufficient. I've found [kilobyte-long compiled regular expressions](https://stackoverflow.com/questions/161738/what-is-the-best-regular-expression-to-check-if-a-string-is-a-valid-url) that validate URLs against RFCs, but I don't think we want that. :-)